File: Python/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        s_dict = {'(': ')', '{': '}', '[': ']'}
        if len(s) % 2 != 0 :return False
        opens = []
        for i in s:
            if i in s_dict.keys():
                opens.append(i)
            elif opens and (i == s_dict[opens[-1]]):
                opens.pop()
            else:
                return False
        if opens: return False
        return True

File: C++/20. Valid Parentheses.cpp
class Solution {
public:
    bool isValid(string s) 
    {
       stack <char> ans ; 
      
      for (int i=0 ; i<s.size();i++)
      {
         if (ans.empty() or s[i]=='[' or s[i]=='(' or s[i]=='{')
           {
               ans.push(s[i]);
           }
        else if (s[i] == ')')
        {
            if(ans.top()=='(')
            {
                ans.pop();
            }
           else
           {
               ans.push(s[i]);
           }
        }
        else if (s[i] == ']')
        {
            if(ans.top()=='[')
            {
                ans.pop();
            }
           else
           {
               ans.push(s[i]);
           }
        }
        else if (s[i] == '}')
        {
            if(ans.top()=='{')
            {
                ans.pop();
            }
           else
           {
               ans.push(s[i]);
           }
        }

      }
    return (ans.empty());
    }
}; 
